test/Spaces/terrain_warp.jl

```diff
@@ -0,0 +1,81 @@
+using Test
+
+import ClimaCore:
+    ClimaCore,
+    Domains,
+    Geometry,
+    Fields,
+    Meshes,
+    Spaces,
+    Topologies,
+    Hypsography
+
+function warp_noop(coord)
+    return zero(coord.x)
+end
+
+function warp_agnesi_peak(coord; a = 100)
+    x = Geometry.component(coord, 1)
+    8 * a^3 / (x^2 + 4 * a^2)
+end
+
+function hvspace_2D(
+    FT = Float64,
+    xlim = (-π, π),
+    zlim = (0, 4π),
+    helem = 10,
+    velem = 50,
+    npoly = 4;
+    stretch = Meshes.Uniform(),
+    warp_fn = warp_agnesi_peak,
+)
+    vertdomain = Domains.IntervalDomain(
+        Geometry.ZPoint{FT}(zlim[1]),
+        Geometry.ZPoint{FT}(zlim[2]);
+        boundary_tags = (:bottom, :top),
+    )
+    vertmesh = Meshes.IntervalMesh(vertdomain, stretch, nelems = velem)
+    vert_face_space = Spaces.FaceFiniteDifferenceSpace(vertmesh)
+
+    horzdomain = Domains.IntervalDomain(
+        Geometry.XPoint{FT}(xlim[1]),
+        Geometry.XPoint{FT}(xlim[2]);
+        periodic = true,
+    )
+    horzmesh = Meshes.IntervalMesh(horzdomain; nelems = helem)
+    horztopology = Topologies.IntervalTopology(horzmesh)
+    quad = Spaces.Quadratures.GLL{npoly + 1}()
+    hspace = Spaces.SpectralElementSpace1D(horztopology, quad)
+
+    z_surface = warp_fn.(Fields.coordinate_field(hspace))
+    f_space = Spaces.ExtrudedFiniteDifferenceSpace(
+        hspace,
+        vert_face_space,
+        Hypsography.LinearAdaption(),
+        z_surface,
+    )
+    c_space = Spaces.CenterExtrudedFiniteDifferenceSpace(f_space)
+
+    return (c_space, f_space)
+end
+
+@testset "2D Extruded Terrain Warped Space" begin
+    for FT in (Float64,)
+        hv_center_space, hv_face_space = hvspace_2D(
+            FT,
+            (-500, 500),
+            (0, 1000),
+            10,
+            20,
+            4;
+            stretch = Meshes.Uniform(),
+            warp_fn = warp_agnesi_peak,
+        )
+        #TODO: tests that the surface coordinate field is correctly computed
```

Comment:
Comparison of `exact z` based on known (discrete) `x` values to `generated` z values extracted from `level(ClimaCore.Fields.(z), 1)` perhaps ? 

Current example with `LinearAdaption` results in the following plot for the generated vs exact surfaces 

<img width="500" alt="Boundary_facecoord_warped" src="https://user-images.githubusercontent.com/43710045/157313017-c922027a-9e8b-4383-9c23-f863b3f3eb09.png">

Reply:
I'll extend this to a resolution sweep if you think this is sufficient and appropriate for this specific function @jakebolewski (e.g. check "generated volume" in domain with varied resolution); this won't address the questions about DSS or horizontal derivative metrics in any case.

Reply:
:+1: , I had a similar test but was unsure what exactly we wanted to do here.

We could add the metric terms test in a similar way I think given we should know the analytical deriv. but  that can come after but I'm pretty sure should correct in any case.

Reply:
You could we take the gradient of the `z` coordinate field, and check that it is aligned correctly?